File: packages/core/src/collection.ts
import type { CollectionName, Key } from "./key";
import type { DeleteOptions, ReadOptions, ScanOptions, WriteOptions } from "./options";
import type { Store } from "./store";

/** Typed handle for values stored in one yugendb collection. */
export class Collection<T> {
  /** Collection name used by this handle. */
  readonly name: CollectionName;
  private readonly store: Store;

  constructor(store: Store, name: CollectionName) {
    this.store = store;
    this.name = name;
  }

  /** Reads a typed value from this collection. */
  get(key: string | Key, options: ReadOptions = {}): Promise<T | null> {
    return this.store.getFromCollection<T>(this.name, key, options);
  }

  /** Writes a typed value to this collection. */
  set(key: string | Key, value: T, options: WriteOptions = {}): Promise<void> {
    return this.store.setInCollection(this.name, key, value, options);
  }

  /** Deletes a value from this collection. */
  delete(key: string | Key, options: DeleteOptions = {}): Promise<boolean> {
    return this.store.deleteFromCollection(this.name, key, options);
  }

  /** Returns whether a non-expired value exists in this collection. */
  exists(key: string | Key): Promise<boolean> {
    return this.store.existsInCollection(this.name, key);
  }

  /** Scans typed values in this collection by key prefix. */
  scanPrefix(
    prefix: string,
    options: Omit<ScanOptions, "prefix"> = {},
  ): Promise<Array<[Key, T]>> {
    return this.store.scanPrefixInCollection<T>(this.name, prefix, options);
  }
}

File: crates/yugendb-core/src/collection.rs
//! Typed collection handle.

use std::marker::PhantomData;

use serde::{de::DeserializeOwned, Serialize};

use crate::{CollectionName, Key, Result, Store, YugenDbError};

/// Typed handle for records in a collection.
#[derive(Clone)]
pub struct Collection<T> {
    store: Store,
    name: CollectionName,
    marker: PhantomData<T>,
}

impl<T> Collection<T> {
    /// Creates a collection handle.
    #[must_use]
    pub fn new(store: Store, name: CollectionName) -> Self {
        Self {
            store,
            name,
            marker: PhantomData,
        }
    }

    /// Returns the collection name.
    #[must_use]
    pub fn name(&self) -> &CollectionName {
        &self.name
    }

    /// Reads a typed value.
    pub async fn get<K>(&self, key: K) -> Result<Option<T>>
    where
        T: DeserializeOwned,
        K: TryInto<Key>,
        YugenDbError: From<K::Error>,
    {
        self.store.get_from_collection(&self.name, key).await
    }

    /// Writes a typed value.
    pub async fn set<K>(&self, key: K, value: &T) -> Result<()>
    where
        T: Serialize,
        K: TryInto<Key>,
        YugenDbError: From<K::Error>,
    {
        self.store.set_in_collection(&self.name, key, value).await
    }

    /// Deletes a value.
    pub async fn delete<K>(&self, key: K) -> Result<bool>
    where
        K: TryInto<Key>,
        YugenDbError: From<K::Error>,
    {
        self.store.delete_from_collection(&self.name, key).await
    }

    /// Returns whether a key exists.
    pub async fn exists<K>(&self, key: K) -> Result<bool>
    where
        K: TryInto<Key>,
        YugenDbError: From<K::Error>,
    {
        self.store.exists_in_collection(&self.name, key).await
    }

    /// Scans typed values by key prefix.
    pub async fn scan_prefix(&self, prefix: impl Into<String>) -> Result<Vec<(Key, T)>>
    where
        T: DeserializeOwned,
    {
        self.store
            .scan_prefix_in_collection(&self.name, prefix)
            .await
    }
}
